spikeinterface/core/addtemplatesrecording.py

```diff
@@ -0,0 +1,161 @@
+    templates: np.ndarray[n_units, n_samples, n_channels]
+        Array containing the templates to inject for all the units.
+    nbefore: list[int] | int | None
+        Where is the center of the template for each unit?
+        If None, will default to the highest peak.
+    amplitude_factor: list[list[float]] | list[float] | float
+        The amplitude of each spike for each unit (1.0=default).
+        Can be sent as a list[float] the same size as the spike vector.
+        Will default to 1.0 everywhere.
+    parent_recording: BaseRecording | None
+        The recording over which to add the templates.
+        If None, will default to traces containing all 0.
+    t_max: list[int] | None
+        The number of frames in the recording.
+
+    Returns
+    -------
+    add_templates_recording: AddTemplatesRecording
+        The recording with the templates added.
+    """
+
+    def __init__(self, sorting: BaseSorting, templates: np.ndarray, nbefore: Union[List[int], int, None] = None,
+                 amplitude_factor: Union[List[List[float]], List[float], float] = 1.0,
+                 parent_recording: Union[BaseRecording, None] = None, t_max: Union[List[int], None] = None) -> None:
+        self._check_templates(templates)
+
+        channel_ids = parent_recording.channel_ids if parent_recording is not None else list(range(templates.shape[2]))
+        BaseRecording.__init__(self, sorting.get_sampling_frequency(), channel_ids, templates.dtype)
+        
+        n_units = len(sorting.unit_ids)
+        assert len(templates) == n_units
+        self.spike_vector = sorting.to_spike_vector()
+
+        if nbefore is None:
+            nbefore = np.argmax(np.max(np.abs(templates), axis=2), axis=1)
+        elif type(nbefore) == int:
+            nbefore = [nbefore]*n_units 
+        else:
+            assert len(nbefore) == n_units
+
+        if type(amplitude_factor) is float:
+            amplitude_factor = np.array([1.0]*len(self.spike_vector), dtype=np.float32)
+        elif len(amplitude_factor) != len(self.spike_vector):  # In this case, it's a list of list for amplitude by unit by spike.
+            tmp = np.array([], dtype=np.float32)
+
+            for segment_index in range(sorting.get_num_segments()):
+                spike_times = [sorting.get_unit_spike_train(unit_id) for unit_id in sorting.unit_ids]
+                spike_times = np.concatenate(spike_times)
+                spike_amplitudes = np.concatenate(amplitude_factor)
+
+                order = np.argsort(spike_times)
+                tmp = np.append(tmp, spike_amplitudes[order])
+
+            amplitude_factor = tmp
+
+        if parent_recording is not None:
+            assert parent_recording.get_num_segments() == sorting.get_num_segments()
+            assert parent_recording.get_sampling_frequency() == sorting.get_sampling_frequency()
+            assert parent_recording.get_num_channels() == templates.shape[2]
```

Comment:
Should we allow for sparse templates here? Or maybe later :)

Reply:
What do you mean by sparse templates?

Reply:
templates defined on a subset of channels, e.g. using the `si.postprocesisng.get_template_channel_sparsity()` function

Reply:
but let's keep this for later ;)